diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
index 393d3cb..2898861 100644
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
@@ -11,6 +11,12 @@
 	compatible = "phicomm,n1", "amlogic,s905d", "amlogic,meson-gxl";
 	model = "Phicomm N1";
 
+	reserved-memory {
+		linux,cma {
+			size = <0x0 0x38000000>;
+		};
+	};
+
 	cvbs-connector {
 		status = "disabled";
 	};
@@ -33,3 +39,36 @@
 &usb {
 	dr_mode = "host";
 };
+
+&sd_emmc_b {
+	/* disable SD CARD */
+	status = "disabled";
+};
+
+&external_phy {
+	/delete-property/ eee-broken-1000t;
+	/delete-property/ max-speed;
+	interrupts = <25 IRQ_TYPE_LEVEL_LOW>;
+};
+
+&ethmac {
+	snps,aal;
+	snps,txpbl = <0x8>;
+	snps,rxpbl = <0x8>;
+};
+
+/* This UART is connected to the Bluetooth module */
+&uart_A {
+	status = "okay";
+	pinctrl-0 = <&uart_a_pins>, <&uart_a_cts_rts_pins>;
+	pinctrl-names = "default";
+	uart-has-rtscts;
+
+	bluetooth {
+		compatible = "brcm,bcm43438-bt";
+		shutdown-gpios = <&gpio GPIOX_17 GPIO_ACTIVE_HIGH>;
+		max-speed = <2000000>;
+		clocks = <&wifi32k>;
+		clock-names = "lpo";
+	};
+};
