diff --git a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
index b95673002..08fe5815e 100644
--- a/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-gxl-s905d-phicomm-n1.dts
@@ -11,6 +11,12 @@ / {
 	compatible = "phicomm,n1", "amlogic,s905d", "amlogic,meson-gxl";
 	model = "Phicomm N1";
 
+    reserved-memory {
+        linux,cma {
+            size = <0x0 0x38000000>;
+        };
+    };
+
 	cvbs-connector {
 		status = "disabled";
 	};
